lcls_tools/common/devices/screen/screen.py

```diff
@@ -0,0 +1,273 @@
+        return self.controls_information.PVs.n_bits.get()
+
+    @property
+    def resolution(self):
+        """The conversion factor of pixels to mm"""
+        return self.controls_information.PVs.resolution.get()
+
+    @property
+    def last_save_filepath(self):
+        """Location and filename for the last file saved by this screen (set in save_images())"""
+        return self._last_save_filepath
+
+    def _generate_new_filename(self, extension: Optional[str] = ".h5") -> str:
+        """
+        Make a new filename for the HDF5 image file
+        Should be of the form: <save-location>/<timestamp>_<screen_name>.h5
+        """
+        stamp = datetime.datetime.now().isoformat()
+        stamp_str = stamp.replace(".", "_").replace("-", "_").replace(":", "_")
+        filename = stamp_str + "_" + self.name + extension
+        path = str(os.path.join(self._root_hdf5_location, filename))
+        return path
+
+    def save_images(
+        self,
+        num_to_capture: int = 1,
+        extra_metadata: Optional[Dict[str, Any]] = None,
+        threaded=True,
+        timeout_in_seconds: int = 10,
+    ):
+        """
+        Collect and saves images to HDF5.
+        Option for threading which spawns a child process
+        in a way that dependant GUIs/Code do not hang.
+        The extra_metadata dictionary will be attached
+        to the metadata for each image in HDF5 .
+
+        """
+        if threaded:
+            work = Thread(
+                target=self._take_images,
+                args=[
+                    num_to_capture,
+                    extra_metadata,
+                    timeout_in_seconds,
+                ],
+            )
+            # normally we join after start, but that blocked the pyqt main thread
+            # so we do not join here. If it breaks, look here first..
+            work.start()
+        else:
+            self._take_images(
+                num_collect=num_to_capture,
+                extra_metadata=extra_metadata,
+                timeout=timeout_in_seconds,
+            )
+
+    def _take_images(
+        self,
+        num_collect: int = 1,
+        extra_metadata: Optional[Dict[str, Any]] = None,
+        timeout: Optional[int] = 20,
+    ):
+        """
+        Performs the work for collecting images and saving to HDF5 file.
+        This procedure will collect all images and then save them all to file.
+
+        If, for any image, we cannot collect within the time provided as timeout,
+        we will stop collecting and save out what was collected before the failure.
+        """
+        self._saving_images = True
+        filename = self._generate_new_filename()
+        captures = []
+        last_updated_at = self.image_timestamp
+        acquisition_start = datetime.datetime.now()
+        while len(captures) != num_collect:
+            if datetime.datetime.now() - acquisition_start > datetime.timedelta(
```

Comment:
This 'timeout' stuff should be moved into a generic util that can be used by other classes as we will want a timeout-functionality for any process we expect to happen in a given time.